src/log/log_slot.c

```diff
@@ -264,21 +267,30 @@ __log_slot_new(WT_SESSION_IMPL *session)
 		}
 		/*
 		 * If we didn't find any free slots signal the worker thread.
+		 * Release the lock so that any threads waiting for it can
+		 * acquire and possibly move things forward.
 		 */
 		WT_STAT_CONN_INCR(session, log_slot_no_free_slots);
 		__wt_cond_signal(session, conn->log_wrlsn_cond);
-		__wt_yield();
+		++wait_count;
+		if (wait_count < WT_THOUSAND)
+			__wt_yield();
+		else {
+			__wt_spin_unlock(session, &log->log_slot_lock);
+			__wt_sleep(0, WT_THOUSAND);
+			__wt_spin_lock(session, &log->log_slot_lock);
+			goto retry;
```

Comment:
Using the goto means the wait_count will be reset after each 1000 spins - I don't see how it could ever reach a million now - which is the condition to timeout. Could this be refactored to avoid a goto?

Reply:
I will refactor and I will yield instead of sleep to try to mitigate some of this.